packages/vitest/src/utils/coverage.ts

```diff
@@ -1,120 +1,194 @@
+        && thresholds.statements === undefined)
+        continue
+
+      // Construct list of coverage summaries where thresholds are compared against
+      const summaries = perFile
+        ? coverageMap.files()
+          .map((file: string) => ({
+            file,
+            summary: coverageMap.fileCoverageFor(file).toSummary(),
+          }))
+        : [{
+            file: null,
+            summary: coverageMap.getCoverageSummary(),
+          }]
+
+      // Check thresholds of each summary
+      for (const { summary, file } of summaries) {
+        for (const thresholdKey of ['lines', 'functions', 'statements', 'branches'] as const) {
+          const threshold = thresholds[thresholdKey]
+
+          if (threshold !== undefined) {
+            const coverage = summary.data[thresholdKey].pct
+
+            if (coverage < threshold) {
+              process.exitCode = 1
+
+              /*
+           * Generate error message based on perFile flag:
+           * - ERROR: Coverage for statements (33.33%) does not meet threshold (85%) for src/math.ts
+           * - ERROR: Coverage for statements (50%) does not meet global threshold (85%)
+           */
+              let errorMessage = `ERROR: Coverage for ${thresholdKey} (${coverage}%) does not meet ${name === GLOBAL_THRESHOLDS_KEY ? name : `"${name}"`} threshold (${threshold}%)`
+
+              if (perFile && file)
+                errorMessage += ` for ${relative('./', file).replace(/\\/g, '/')}`
+
+              console.error(errorMessage)
+            }
           }
         }
       }
     }
   }
 
+  /**
+   * TODO describe this
+   */
+  resolveThresholds({ coverageMap, thresholds, createCoverageMap }: {
+    coverageMap: CoverageMap
+    thresholds: NonNullable<BaseCoverageOptions['thresholds']>
+    createCoverageMap: () => CoverageMap
+  }): ResolvedThreshold[] {
+    const resolvedThresholds: ResolvedThreshold[] = []
+    const files = coverageMap.files()
+    const filesMatchedByGlobs: string[] = []
+    const globalCoverageMap = createCoverageMap()
+
+    for (const key of Object.keys(thresholds) as (`${keyof typeof thresholds}`[])) {
+      if (key === 'perFile' || key === 'autoUpdate' || key === '100' || THRESHOLD_KEYS.includes(key))
+        continue
+
+      const glob = key
+      const globThresholds = resolveGlobThresholds(thresholds[glob])
+      const globCoverageMap = createCoverageMap()
+
+      const matchingFiles = files.filter(file => mm.isMatch(file, glob))
+      filesMatchedByGlobs.push(...matchingFiles)
+
+      for (const file of matchingFiles) {
+        const fileCoverage = coverageMap.fileCoverageFor(file)
+        globCoverageMap.addFileCoverage(fileCoverage)
+      }
+
+      resolvedThresholds.push({
+        name: glob,
+        coverageMap: globCoverageMap,
+        thresholds: globThresholds,
+      })
+    }
+
+    // TODO: Global threshold should include all files, even the ones defined for specific globs.
+    // Remove filter.
```

Comment:
Note for future references:

This is actually working correctly now. Files included in their own glob thresholds should not be counted into global coverage when checking thresholds. When using `text` reporter it might cause some confusion as coverage percentages do not match thresholds configuration. 

Below is example from Jest where `text` reporter reports 33% coverage for branches, but custom threshold configuration moves one file into its own set and lowers global branch coverage:

![image](https://github.com/vitest-dev/vitest/assets/14806298/4eb753e0-dbba-4273-bada-b1f297b5a6ce)